src/_pytest/mark/__init__.py

```diff
@@ -211,24 +215,41 @@ def deselect_by_keyword(items: "List[Item]", config: Config) -> None:
         items[:] = remaining
 
 
+NOT_NONE_SENTINEL = object()
+
+
 @dataclasses.dataclass
 class MarkMatcher:
     """A matcher for markers which are present.
 
     Tries to match on any marker names, attached to the given colitem.
     """
 
-    __slots__ = ("own_mark_names",)
+    __slots__ = ("own_mark_name_mapping",)
 
-    own_mark_names: AbstractSet[str]
+    own_mark_name_mapping: Dict[str, List["Mark"]]
 
     @classmethod
     def from_item(cls, item: "Item") -> "MarkMatcher":
-        mark_names = {mark.name for mark in item.iter_markers()}
-        return cls(mark_names)
+        mark_name_mapping = collections.defaultdict(list)
+        for mark in item.iter_markers():
+            mark_name_mapping[mark.name].append(mark)
+        return cls(mark_name_mapping)
+
+    def __call__(self, name: str, /, **kwargs: object) -> bool:
+        if not (matches := self.own_mark_name_mapping.get(name, [])):
+            return False
+
+        if not kwargs:
+            return True
 
-    def __call__(self, name: str) -> bool:
-        return name in self.own_mark_names
+        for mark in matches:
+            if all(
+                mark.kwargs.get(k, NOT_NONE_SENTINEL) == v for k, v in kwargs.items()
```

Comment:
Sentinel objects are typically compared though the identity operator, not equality (`is None`, `is not False` etc.)

So the following might be more correct:
```suggestion
                mark.kwargs.get(k, NOT_NONE_SENTINEL) is v for k, v in kwargs.items()
```

Reply:
Or perhaps even
```suggestion
                v is NOT_NONE_SENTINEL or mark.kwargs.get(k, NOT_NONE_SENTINEL) == v for k, v in kwargs.items()
```

Reply:
I do not really care about the sentinel object, I just need it to not screw up comparing e.g. `None` (when I did not use any). The other values I need to compare have to be compared by value.